### DISCUSSION TITLE
Can I create separate local instances for the test suite and for development?

### DISCUSSION THREAD

**Discussion by jtlapp** [supabase/supabase#14329] [Questions] [answered]
I need two local instances to do development:

1. One instance for the test suite, which tears everything down between tests.
2. The other instance is for interactive development, so I can play with the website and through the website change the contents of the database. This is my hot-reloading site.

The two databases need not exist at the same time, but they each need Supabase's auth support tables, and I need to preserve data in the second instance.

It seems to be possible to do this via two projects. Is that the only way to do this?

To keep it to one project, would I have to run the instances on separate computers?

**Comment by Olyno** [comment, marked as answer]
Hi 👋🏻 

You can totally achieve this within a single project by leveraging Docker and docker-compose. What you can do is set up two separate docker-compose configurations for your test and development environments. Each configuration would spin up its own Postgres database and Supabase services with isolated data.

When you want to run your test suite, you would use the test docker-compose setup, and for interactive development, you'd use the development docker-compose setup. This way, both environments would have their own auth support tables and data.

Remember, you can manipulate the ``SUPABASE_JWT_SECRET`` and ``SUPABASE_SERVICE_ROLE_KEY`` in your ``.env`` file to have different values for your test and development environments. This helps to maintain the isolation you're seeking.

You don't need two separate computers, just two separate docker-compose configurations. You just need to make sure the ports for your test and development instances don't collide.

This setup will let you spin up or tear down each environment as needed without affecting the other. And you get to keep all the Supabase goodness in both places!

**Reply by jtlapp** [reply]
Thank you so much! I'll see what I can do. I've been meaning to learn some docker, anyway.

**Comment by jtlapp** [comment]
I found an easy way to do this in my monorepo. I created separate packages for `test-instance` and `dev-instance` and installed `supabase` in each, initializing each with its own `.toml`. I changed the ports in the test instance. And now to run either instance I use one of these commands:

```
    "start-dev-instance": "(cd instances/dev-instance && npx supabase start)",
    "stop-dev-instance": "(cd instances/dev-instance && npx supabase stop)",
    "start-test-instance": "(cd instances/test-instance && npx supabase start)",
    "stop-test-instance": "(cd instances/test-instance && npx supabase stop)",
```

I'm using Next.js, which loads environment variables from `.env`. My unit tests currently manually load from `.env.test`, so I expect to have to figure out how to get Next.js to use the proper `.env` file once I get to integration tests. But one problem at a time...

**Reply by amerryma** [reply]
Have you solidified this strategy out any further? I'm looking for something similar. How do you solve the problem where each supabase folder essentially holds their own migrations folders? Symlinks?

**Reply by jtlapp** [reply]
I'm using the same migrations folder for both, but at least in my case, it would be easy for me to separate them. I'm using [Kysely migrations](https://kysely.dev/docs/migrations), which requires me to write some of my own wrapper code. I had that wrapper code the path to the migrations file. I don't rely on Supabase to install my app schema.

**Reply by amerryma** [reply]
Thanks for the reply! It makes sense, considering if you self host Supabase it doesn't seem like there is a tool for managing migrations either.

**Reply by jtlapp** [reply]
Well, I'm self hosting for now, just for development. I do plan to use the hosting service. Don't know what trouble I'll run into then.
